datastore/providers/postgres_datastore.py

```diff
@@ -0,0 +1,332 @@
+        logging.info("Creating table %s" % POSTGRES_TABLENAME)
+        create_table = (
+            "CREATE TABLE IF NOT EXISTS %s (doc_id TEXT, chunk_id TEXT, text TEXT, embedding vector(%s), metadata JSONB)"
+            % (POSTGRES_TABLENAME, str(VECTOR_DIMENSION))
+        )
+        cur = conn.cursor()
+        cur.execute(create_table)
+        index_exists_staetment = (
+            "SELECT EXISTS (SELECT 1 FROM pg_indexes  WHERE tablename = '%s' AND indexname = '%s')"
+            % (POSTGRES_TABLENAME, POSTGRES_DOCID_INDEX)
+        )
+        cur.execute(index_exists_staetment)
+        self.docid_index_exists = cur.fetchall()[0][0]
+
+        if not self.docid_index_exists:
+            index_doc_ids = (
+                "CREATE INDEX CONCURRENTLY ON %s (doc_id) " % POSTGRES_TABLENAME
+            )
+            cur.execute(index_doc_ids)
+
+        self.index_counter = 0
+        index_exists_staetment = (
+            "SELECT EXISTS (SELECT 1 FROM pg_indexes  WHERE tablename = '%s' AND indexname = '%s')"
+            % (POSTGRES_TABLENAME, POSTGRES_EMBEDDING_INDEX)
+        )
+        cur.execute(index_exists_staetment)
+        self.index_exists = cur.fetchall()[0][0]
+
+        cur.close()
+        self.pool.putconn(conn)
+
+    async def _upsert(self, chunks: Dict[str, List[DocumentChunk]]) -> List[str]:
+        """
+        Takes in a dict from document id to list of document chunks and inserts them into the index.
+        Return a list of document ids.
+        """
+        # Create a connection
+
+        conn = self.pool.getconn()
+        conn.autocommit = True
+        cur = conn.cursor()
+
+        # Initialize a list of ids to return
+        doc_ids: List[str] = []
+        if POSTGRES_UPSERT_COMMAND == "INSERT":
+            # Loop through docs/chunks
+            for doc_id, chunk_list in chunks.items():
+                doc_ids.append(doc_id)
+                for chunk in chunk_list:
+                    metadata = chunk.metadata.dict()
+                    if "created_at" in list(metadata.keys()):
+                        if metadata["created_at"]:
+                            metadata["created_at"] = to_unix_timestamp(
+                                metadata["created_at"]
+                            )
+                    metadata = json.dumps(metadata)
+                    insert_statement = sql.SQL(
+                        "INSERT INTO {} (doc_id, chunk_id, text, embedding, metadata) VALUES ({}, {}, {}, {}, {})"
+                    ).format(
+                        sql.Identifier(POSTGRES_TABLENAME),
+                        sql.Literal(doc_id),
+                        sql.Literal(chunk.id),
+                        sql.Literal(chunk.text),
+                        sql.Literal(chunk.embedding),
+                        sql.Literal(metadata),
+                    )
+                    cur.execute(insert_statement)
+                    self.index_counter += 1
```

Comment:
You need to clear any no longer present chunk ids as well for upserting a doc, pseudocode.

`delete from POSTGRES_TABLENAME WHERE doc_id = X and chunk_id not in ({[chunk.id for chunk in chunks ]})`